### PR TITLE
ARROW-14171: [C++][Python][Packaging] Upgrade VCPKG version and add google-cloud-cpp dependency

### PR DESCRIPTION
**Pull request by coryan**

### CODE REVIEW

**Comment by coryan** [comment]
Ping? Is there something else I need to do here?

**Comment by kou** [comment]
It seems that this is a problem.

https://github.com/ursacomputing/crossbow/runs/4039103923?check_suite_focus=true#step:6:1189

```text
#12 14.69 Package google-cloud-cpp does not have a storage feature
```

**Comment by coryan** [comment]
I see.  The code checks out version `2021.04.30`:

https://github.com/apache/arrow/blob/faf8b827493afee60ea44893b64d36a7d5f3fadc/.env#L75

That version does not include the `storage` feature, neither does the latest tag at `2021.05.12`.

How do you feel about pinning vcpkg to a specific SHA, but not to a tag?

**Comment by kou** [comment]
> How do you feel about pinning vcpkg to a specific SHA, but not to a tag?

@ianmcook @kszucs What do you think about this?

**Comment by kszucs** [comment]
We used to pin specific SHAs beforehand, so it's fine.

**Comment by coryan** [comment]
Thanks, I have updated this PR accordingly.  I also sent ursacomputing/crossbow#6 for review, I am not sure which one should be merged first, please let me know if I should help with that.

**Comment by kszucs** [comment]
Thanks! Since it affects the build cache we should wait for the 6.0.1 patch release with both of the PRs.

**Comment by kszucs** [comment]
Merged the crossbow PR. I'm going to rebuild and push the windows docker images for the windows wheels then we can submit the crossbow jobs here.

**Comment by kszucs** [comment]
After fixing the `vcpkg/ports.patch` I'm getting a vcpkg dependency problem on windows:

```
✖ REPO=ghcr.io/ursacomputing/arrow PYTHON=3.10 archery docker build --using-docker-cli python-wheel-windows-vs2017
Error response from daemon: manifest unknown
docker pull ghcr.io/ursacomputing/arrow:python-3.10-wheel-windows-vs2017-vcpkg-9675ab15def74a0b96d954bddc7ddabb42c92452 exited with non-zero exit code 1
Sending build context to Docker daemon  453.1kB
Step 1/19 : FROM abrarov/msvc-2017:2.10.0
 ---> 48bb8164f545
Step 2/19 : RUN choco install --no-progress -r -y cmake --installargs 'ADD_CMAKE_TO_PATH=System' &&     choco install --no-progress -r -y gzip wget ninja
 ---> Using cache
 ---> 84da6558371c
Step 3/19 : RUN setx path "%path%;C:\Program Files\Git\usr\bin"
 ---> Using cache
 ---> 744f355e6df4
Step 4/19 : ARG vcpkg
 ---> Using cache
 ---> 6fd7cae10c08
Step 5/19 : RUN git clone https://github.com/Microsoft/vcpkg &&     vcpkg\bootstrap-vcpkg.bat -disableMetrics &&     setx PATH "%PATH%;C:\vcpkg" &&     git -C vcpkg checkout %vcpkg%
 ---> Using cache
 ---> a85f4b3090f6
Step 6/19 : COPY ci/vcpkg/*.patch      ci/vcpkg/*windows*.cmake      arrow/ci/vcpkg/
 ---> Using cache
 ---> cdc9fbfb7e6e
Step 7/19 : RUN cd vcpkg && git apply --ignore-whitespace C:/arrow/ci/vcpkg/ports.patch
 ---> Using cache
 ---> c93972ad92dc
Step 8/19 : ARG build_type=release
 ---> Using cache
 ---> c130e80c70b8
Step 9/19 : ENV CMAKE_BUILD_TYPE=${build_type}     VCPKG_OVERLAY_TRIPLETS=C:\\arrow\\ci\\vcpkg     VCPKG_DEFAULT_TRIPLET=x64-windows-static-md-${build_type}     VCPKG_FEATURE_FLAGS=-manifests
 ---> Using cache
 ---> e7a18a9e6535
Step 10/19 : RUN vcpkg install --clean-after-build         abseil         aws-sdk-cpp[config,cognito-identity,core,identity-management,s3,sts,transfer]         boost-filesystem         boost-multiprecision         boost-system         brotli         bzip2         c-ares         curl         flatbuffers         gflags         glog         google-cloud-cpp[core,storage]         grpc         lz4         openssl         orc         protobuf         rapidjson         re2         snappy         thrift         utf8proc         zlib         zstd
 ---> Running in 3c89420370c9
Computing installation plan...
Error: aws-checksums[core] is only supported on '!arm & !windows'
The command 'cmd /S /C vcpkg install --clean-after-build         abseil         aws-sdk-cpp[config,cognito-identity,core,identity-management,s3,sts,transfer]         boost-filesystem         boost-multiprecision         boost-system         brotli         bzip2         c-ares         curl         flatbuffers         gflags         glog         google-cloud-cpp[core,storage]         grpc         lz4         openssl         orc         protobuf         rapidjson         re2         snappy         thrift         utf8proc         zlib         zstd' returned a non-zero code: 1
Error: docker build --build-arg BUILDKIT_INLINE_CACHE=1 --build-arg python=3.10 --build-arg vcpkg=9675ab15def74a0b96d954bddc7ddabb42c92452 -f ci/docker/python-wheel-windows-vs2017.dockerfile -t ghcr.io/ursacomputing/arrow:python-3.10-wheel-windows-vs2017-vcpkg-9675ab15def74a0b96d954bddc7ddabb42c92452 /Users/kszucs/Workspace/arrow exited with non-zero exit code 1
```

It has been already resolved upstream, so I pinned the HEAD commit.

**Comment by kszucs** [comment]
@github-actions crossbow submit -g wheel

**Comment by kszucs** [comment]
New changes:
- bumped the vcpkg version to make the aws-sdk-cpp compile on windows
- updated to use the `install_vcpkg.sh` script everywhere (including the windows docker images)
- removed the duplicated vcpkg versions across the build configs, now it uses VCPKG version from `.env`

Still, We have several problems here:
1. CMake `3.22.0` has a regression causing issue with [AWSSDKConfig.cmake](https://github.com/ursacomputing/crossbow/runs/4313032404?check_suite_focus=true#step:11:282). We need to pin `3.21.4` both on macOS and manylinux.
2. Manylinux2010 [doesn't have `realpath`](https://github.com/ursacomputing/crossbow/runs/4313036902?check_suite_focus=true#step:7:639) so we need to switch to `readlink -f` in the `install_vcpkg.sh` script.
3. The vcpkg installer now downloads a prebuilt binary instead of compiling the vcpkg tool itself locally but the binary links with `>glibc 2.14` and the `manylinux2010` image has `glibc=2.12`. I patched the `bootstrap.sh` script to force build `vcpkg` but getting many ambiguous calls errors during compilation. Perhaps it would be the right time to drop the EOL manylinux2010 packages.
4. The aws-sdk-cpp packages [are restricted to non-arm architectures](https://github.com/ursacomputing/crossbow/runs/4313053277?check_suite_focus=true#step:8:42) despite that earlier versions used to work. I assume patching the port files with a new constraint would work.

**Comment by kszucs** [comment]
@github-actions crossbow submit -g wheel

**Comment by kszucs** [comment]
@github-actions crossbow submit wheel-macos-big-sur-*

**Comment by kszucs** [comment]
@github-actions crossbow submit wheel-macos-big-sur-*

**Comment by coryan** [comment]
@kszucs thanks for all the work in this PR.  Is there something I could do now (or could have done before) to make your life easier?

**Comment by kszucs** [comment]
@github-actions crossbow submit wheel-macos-big-sur-*

**Comment by kszucs** [comment]
@github-actions crossbow submit wheel-macos-mavericks-cp38-amd64

**Comment by kszucs** [comment]
@github-actions crossbow submit wheel-macos-big-sur-cp39-arm64

**Comment by kszucs** [comment]
@coryan seems like google-cloud-cpp fails to compile on macOS: https://github.com/ursacomputing/crossbow/runs/4326265101?check_suite_focus=true#step:9:1711

**Comment by kszucs** [comment]
@github-actions crossbow submit -g wheel

**Comment by coryan** [comment]
> @coryan seems like google-cloud-cpp fails to compile on macOS: https://github.com/ursacomputing/crossbow/runs/4326265101?check_suite_focus=true#step:9:1711

Maybe we can just update the vcpkg baseline in this PR, and add `google-cloud-cpp` separately?

As to the build failure, seems like this build is using `clang++`, and I think both we and the vcpkg folks test with `c++` (Apple's native compiler). I wonder if that makes a difference.

**Comment by kszucs** [comment]
It cannot find ssl/crypto:

```
CMake Error at /usr/local/Cellar/cmake/3.22.0/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find OpenSSL, try to set the path to OpenSSL root folder in the
  system variable OPENSSL_ROOT_DIR (missing: OPENSSL_CRYPTO_LIBRARY
  OPENSSL_INCLUDE_DIR)
Call Stack (most recent call first):
  /usr/local/Cellar/cmake/3.22.0/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:594 (_FPHSA_FAILURE_MESSAGE)
  /usr/local/Cellar/cmake/3.22.0/share/cmake/Modules/FindOpenSSL.cmake:574 (find_package_handle_standard_args)
  /tmp/picsaba/scripts/buildsystems/vcpkg.cmake:788 (_find_package)
  google/cloud/storage/CMakeLists.txt:49 (find_package)
```

Installing openssl before google-cloud-sdk resolves the issue. I'm going to use that as a workaround but this should be fixed upstream. 

Note, that the following triplet file is used https://github.com/apache/arrow/blob/master/ci/vcpkg/x64-osx-static-release.cmake

**Comment by kszucs** [comment]
@github-actions crossbow submit -g wheel

**Comment by kszucs** [comment]
@github-actions crossbow submit -g wheel

**Comment by kszucs** [comment]
@github-actions crossbow submit wheel-macos-big-sur-*

**Comment by kszucs** [comment]
@github-actions crossbow submit -g wheel

**Comment by kszucs** [comment]
@github-actions crossbow submit wheel-macos-big-sur-*-universal2

File: dev/archery/archery/docker.py [hunk 1]
```diff
@@ -1,402 +0,0 @@
-# Licensed to the Apache Software Foundation (ASF) under one
```
Comment (kszucs) [inline dev/archery/archery/docker.py:1]: This file should have been removed already.

**Comment by kszucs** [comment]
@github-actions crossbow submit wheel-macos-big-sur-*-universal2

File: dev/tasks/python-wheels/github.linux.amd64.yml [hunk 2]
```diff
@@ -47,7 +47,7 @@ jobs:
       {{ macros.github_upload_releases("arrow/python/repaired_wheels/*.whl")|indent }}
       {{ macros.github_upload_gemfury("arrow/python/repaired_wheels/*.whl")|indent }}
 
-      {% if arrow.branch == 'master' %}
+      {% if True or arrow.branch == 'master' %}
```
Comment (kszucs) [inline dev/tasks/python-wheels/github.linux.amd64.yml:50]: Remove before merging.

File: dev/tasks/python-wheels/travis.linux.arm64.yml [hunk 3]
```diff
@@ -66,7 +66,7 @@ after_success:
   {{ macros.travis_upload_releases("arrow/python/repaired_wheels/*.whl") }}
   {{ macros.travis_upload_gemfury("arrow/python/repaired_wheels/*.whl") }}
 
-  {% if arrow.branch == 'master' %}
+  {% if True or arrow.branch == 'master' %}
```
Comment (kszucs) [inline dev/tasks/python-wheels/travis.linux.arm64.yml:69]: Remove before merging.

**Comment by kszucs** [comment]
@github-actions crossbow submit wheel-macos-big-sur-*-universal2

**Comment by kszucs** [comment]
@coryan we don't yet have python bindings for GCS, right?

**Comment by kszucs** [comment]
@github-actions crossbow submit wheel-macos-big-sur-*-universal2

**Comment by kszucs** [comment]
@github-actions crossbow submit wheel-macos-big-sur-*-universal2

**Comment by kszucs** [comment]
@github-actions crossbow submit wheel-macos-big-sur-cp39-universal2

**Comment by kszucs** [comment]
@github-actions crossbow submit -g wheel

**Comment by kszucs** [comment]
@github-actions crossbow submit wheel-macos-big-sur-cp39-universal2

**Comment by kszucs** [comment]
@github-actions crossbow submit wheel-macos-big-sur-cp39-universal2

**Comment by kszucs** [comment]
@github-actions crossbow submit -g wheel

**Comment by coryan** [comment]
> @coryan we don't yet have python bindings for GCS, right?

If you mean in Arrow, then no, they do not exist yet.  I had no plans to create them either.

**Comment by pitrou** [comment]
> If you mean in Arrow, then no, they do not exist yet. I had no plans to create them either.

That's something that other contributors can do :-)

**Comment by kszucs** [comment]
Added a note about enabling the GCS feature in the wheels for https://issues.apache.org/jira/browse/ARROW-14892

@pitrou this should be ready for review then

File: ci/docker/python-wheel-manylinux-201x.dockerfile [hunk 4]
```diff
@@ -18,65 +18,61 @@
...
+ARG glibc=2.18
 COPY ci/vcpkg/*.patch \
      ci/vcpkg/*linux*.cmake \
      arrow/ci/vcpkg/
-RUN cd /opt/vcpkg && git apply --ignore-whitespace /arrow/ci/vcpkg/ports.patch
+COPY ci/scripts/install_vcpkg.sh \
+     ci/scripts/install_glibc.sh \
+     arrow/ci/scripts/
+RUN arrow/ci/scripts/install_vcpkg.sh /opt/vcpkg ${vcpkg} && \
+    if [ "${manylinux}" == "2010" ]; then \
```
Comment (pitrou) [inline ci/docker/python-wheel-manylinux-201x.dockerfile:57]: Can you add a comment explaining this? Also, why put all this on the same `RUN` line as vcpkg?

File: ci/scripts/install_glibc.sh [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+url="http://ftp.gnu.org/gnu/glibc/glibc-${version}.tar.gz"
+
+mkdir /tmp/glibc
+wget -q ${url} -O - | tar -xzf - --directory /tmp/glibc --strip-components=1
+
+mkdir /tmp/glibc/build
+pushd /tmp/glibc/build
+../configure --prefix=${prefix}
+make -j4
```
Comment (pitrou) [inline ci/scripts/install_glibc.sh:38]: Do we *really* need to build the glibc from scratch?

Comment (kszucs) [reply ci/scripts/install_glibc.sh:38]: You mean we could use a precompiled binary?

Comment (pitrou) [reply ci/scripts/install_glibc.sh:38]: I don't know :-) Is the glibc build because of vcpkg compatibility problems?

Comment (kszucs) [reply ci/scripts/install_glibc.sh:38]: Yes, vcpkg-tool (the vcpkg binary) requires glibc>=2.14.

Comment (pitrou) [reply ci/scripts/install_glibc.sh:38]: I see. Apparently building the glibc is faster that I feared. Still, can you make this `make -j$(nproc)`?

File: ci/docker/python-wheel-manylinux-201x.dockerfile [hunk 6]
```diff
@@ -18,65 +18,61 @@
...
 # Need to install the boost-build prior installing the boost packages, otherwise
 # vcpkg will raise an error.
 # TODO(kszucs): factor out the package enumeration to a text file and reuse it
 # from the windows image and potentially in a future macos wheel build
 RUN vcpkg install --clean-after-build \
-        boost-build:${arch_short_alias}-linux && \
-    vcpkg install --clean-after-build \
         abseil \
         aws-sdk-cpp[config,cognito-identity,core,identity-management,s3,sts,transfer] \
         boost-filesystem \
```
Comment (pitrou) [inline ci/docker/python-wheel-manylinux-201x.dockerfile:78]: Hmm, do we still need boost-filesystem here? AFAIR the dependency is only for unit tests.

Comment (kszucs) [reply ci/docker/python-wheel-manylinux-201x.dockerfile:78]: Going to create a follow-up to remove it.

Comment (kszucs) [reply ci/docker/python-wheel-manylinux-201x.dockerfile:78]: https://issues.apache.org/jira/browse/ARROW-14928

File: ci/scripts/install_glibc.sh (hunk 5 above)
Comment (kszucs) [reply ci/scripts/install_glibc.sh]: Fixed and tested locally.

File: ci/docker/python-wheel-manylinux-201x.dockerfile (hunk 4 above)
Comment (pitrou) [reply ci/docker/python-wheel-manylinux-201x.dockerfile:57]: Ah, my bad. Thank you.

**Review by kszucs — APPROVED** [review]

**Pull request closed without merging**